packages/angular/ssr/src/routes/route-tree.ts

```diff
@@ -173,79 +183,122 @@ export class RouteTree {
-   * @param route - The route path to match.
-   * @returns A generator yielding route nodes that match the given route.
-   */
-  private *traverseByRoute(route: string): Generator<RouteTreeNode> {
-    function* traverseNode(
-      node: RouteTreeNode,
-      remainingSegments: string[],
-    ): Generator<RouteTreeNode> {
-      const { metadata, children } = node;
-
-      if (!remainingSegments.length) {
-        // At a leaf node, and metadata if present yield the node
-        if (metadata) {
-          yield node;
-        }
+    // If the node has no children, end the traversal
+    if (!children.size) {
+      return;
+    }
 
-        return;
+    const [segment, ...restSegments] = remainingSegments;
+    let bestMatchNode: RouteTreeNode | undefined;
+
+    // 1. Exact segment match
+    const exactMatchNode = node.children.get(segment);
+    if (exactMatchNode) {
+      const matchingNode = this.traverseBySegments(restSegments, exactMatchNode);
+      if (
+        matchingNode &&
+        (!bestMatchNode || this.isPrioritizedByInsertionIndex(bestMatchNode, matchingNode))
+      ) {
+        bestMatchNode = matchingNode;
       }
+    }
 
-      if (!children.size) {
-        return;
+    // 2. Wildcard segment match (`*`)
+    const wildcardNode = node.children.get('*');
+    if (wildcardNode) {
+      const matchingNode = this.traverseBySegments(restSegments, wildcardNode);
+      if (
+        matchingNode &&
+        (!bestMatchNode || this.isPrioritizedByInsertionIndex(bestMatchNode, matchingNode))
+      ) {
+        bestMatchNode = matchingNode;
       }
+    }
 
-      const [segment, ...restSegments] = remainingSegments;
+    // 3. Deep wildcard segment match (`**`)
+    const deepWildcardNode = node.children.get('**');
+    if (
+      deepWildcardNode &&
+      (!bestMatchNode || this.isPrioritizedByInsertionIndex(bestMatchNode, deepWildcardNode))
+    ) {
+      bestMatchNode = deepWildcardNode;
+    }
 
-      // Attempt to match the exact segment
-      let childNode = children.get(segment);
-      if (childNode) {
-        yield* traverseNode(childNode, restSegments);
-      }
+    return bestMatchNode;
+  }
 
-      // Attempt to match a wildcard segment
-      childNode = children.get('*');
-      if (childNode) {
-        yield* traverseNode(childNode, restSegments);
-      }
-    }
+  /**
+   * Determines if the candidate node is prioritized based on its insertion index compared to the current best match node.
+   * A node with a lower insertion index is prioritized as it was defined earlier.
+   *
+   * @param currentBestMatchNode - The current best match node.
+   * @param candidateNode - The node being evaluated for higher priority based on insertion index.
+   * @returns `true` if the candidate node is prioritized (i.e., has a lower insertion index)
+   * compared to the current best match node, otherwise `false`.
+   */
+  private isPrioritizedByInsertionIndex(
```

Comment:
Consider: Rather than treating the current best match specially, consider treating the parameters symmetrically and just returning the best one. If you check for `null` as well you might be able to simplify the usage to just:

```typescript
bestMatchNode = firstByInsertionIndex(bestMatchNode, node.children.get(segment));
bestMatchNode = firstByInsertionIndex(bestMatchNode, node.children.get('*'));
bestMatchNode = firstByInsertionIndex(bestMatchNode, node.children.get('**'));
```

Alternatively you could write this like a compare function, sort an array of the candidates with it, and then pick the first one.

Reply:
Good idea